timdex_dataset_api/dataset.py

```diff
@@ -503,65 +510,96 @@ def _yield_current_record_batches(
+        try:
+            unfiltered_batch = next(unfiltered_batches)
+        except StopIteration:
+            return
 
-            # check each record id and track unseen ones
-            for i, record_id in enumerate(batch.column("timdex_record_id").to_pylist()):
+        seen_records: set[str] = set()
+        previous_fragment = 0
+        for batch in batches:
+
+            if len(batch) == 0:
+                continue
+
+            # if the fragment has incremented, we've moved to the next fragment so we need
+            # to backfill ids from the last unfiltered fragment as seen
+            current_fragment = get_batch_fragment_index(batch)
+            if current_fragment > previous_fragment:
+                unfiltered_batch, timdex_record_ids = (
+                    self._get_previous_unfiltered_fragment_ids(  # type: ignore[assignment]
+                        current_fragment,
+                        unfiltered_batch,
+                        unfiltered_batches,
+                    )
+                )
+                seen_records.update(timdex_record_ids)
+            previous_fragment = current_fragment
+
+            # find indices of records we haven't seen yet
+            unseen_batch_indices = []
+            batch_record_ids = batch.column("timdex_record_id").to_pylist()
+            for i, record_id in enumerate(batch_record_ids):
                 if record_id not in seen_records:
                     unseen_batch_indices.append(i)
+                    seen_records.add(record_id)  # type: ignore[arg-type]
 
-            # even if not a record to yield, update our list of seen records from all
-            # records in the unfiltered batch
-            seen_records.update(unfiltered_batch.column("timdex_record_id").to_pylist())
-
-            # if no unseen records from this batch, skip yielding entirely
+            # if we don't find any, continue with next batch
+            # else, yield a RecordBatch of the unseen records
             if not unseen_batch_indices:
                 continue
+            yield batch.take(pa.array(unseen_batch_indices))  # type: ignore[arg-type]
 
-            # create a new RecordBatch using the unseen indices of the batch
-            _batch = batch.take(pa.array(unseen_batch_indices))  # type: ignore[arg-type]
-            if len(_batch) > 0:
-                yield _batch
+    def _get_previous_unfiltered_fragment_ids(
+        self,
+        target_fragment: int,
+        batch: pa.RecordBatch,
+        unfiltered_batches: Iterator[pa.RecordBatch],
+    ) -> tuple[pa.RecordBatch | None, set[str]]:
+        """Method to get all timdex_record_ids from an unfiltered fragment.
+
+        Returns:
+            tuple: [next fragment batch, previous fragment ids]
+        """
+        current_fragment = get_batch_fragment_index(batch)
+
+        timdex_record_ids: set[str] = set()
+        while current_fragment < target_fragment:
```

Comment:
This is effectively exhausting all batches in the previous fragment, adding those `timdex_record_id`'s to a list we'll return.